pkgs/stdenv/generic/check-meta.nix

```diff
@@ -372,7 +444,56 @@ let
         }" ]
     else
       [ "key 'meta.${k}' is unrecognized; expected one of: \n  [${concatMapStringsSep ", " (x: "'${x}'") (attrNames metaTypes)}]" ];
-  checkMeta = meta: optionals config.checkMeta (concatMap (attr: checkMetaAttr attr meta.${attr}) (attrNames meta));
+  checkMeta = meta: optionals config.checkMeta (concatMap (attr: checkMetaAttr attr meta.${attr}) (attrNames meta)
+
+    # Extended checks for problems, as they do not fit the module system
+    ++ lib.optionals (meta ? problems) (
+
+      # Check problem kinds are correct
+      lib.pipe meta.problems [
```

Comment:
This all looks very allocation/iteration heavy:
I'd completely avoid using `lib.pipe` in _any_ hot paths for one.

For a normal code review I wouldn't be so pedantic about this, but check-meta is a performance special case.
Changes here needs to be carefully measured, and the Ofborg performance report doesn't include `checkMeta`, so it's entirely useless in this context.

Measuring nixpkgs eval with `checkMeta = true` before and after this PR (`env NIX_SHOW_STATS=1 nix-env -qa -f ./outpaths.nix --arg checkMeta true > /dev/null`):

- Before
``` json
{
  "cpuTime": 271.2768249511719,
  "envs": {
    "bytes": 7850426888,
    "elements": 413454253,
    "number": 283924554
  },
  "gc": {
    "heapSize": 22804639744,
    "totalBytes": 47139879824
  },
  "list": {
    "bytes": 987373624,
    "concats": 21648163,
    "elements": 123421703
  },
  "nrAvoided": 329081776,
  "nrFunctionCalls": 261979419,
  "nrLookups": 131336801,
  "nrOpUpdateValuesCopied": 675865734,
  "nrOpUpdates": 32046977,
  "nrPrimOpCalls": 146549876,
  "nrThunks": 399878601,
  "sets": {
    "bytes": 15552994336,
    "elements": 911935337,
    "number": 60126809
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 16,
    "Value": 24
  },
  "symbols": {
    "bytes": 2887399,
    "number": 178950
  },
  "values": {
    "bytes": 12350285592,
    "number": 514595233
  }
}
```

- After
``` json
{
  "cpuTime": 296.9483947753906,
  "envs": {
    "bytes": 8506454064,
    "elements": 438912624,
    "number": 312197067
  },
  "gc": {
    "heapSize": 22737526784,
    "totalBytes": 48902439376
  },
  "list": {
    "bytes": 1000276576,
    "concats": 22264431,
    "elements": 125034572
  },
  "nrAvoided": 349063893,
  "nrFunctionCalls": 289412278,
  "nrLookups": 149742381,
  "nrOpUpdateValuesCopied": 694111348,
  "nrOpUpdates": 35108342,
  "nrPrimOpCalls": 168992563,
  "nrThunks": 417292743,
  "sets": {
    "bytes": 15871244064,
    "elements": 927562812,
    "number": 64389942
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 16,
    "Value": 24
  },
  "symbols": {
    "bytes": 4092168,
    "number": 234982
  },
  "values": {
    "bytes": 12774278736,
    "number": 532261614
  }
}
```

Reply:
What could be used to have better performance whilst keeping the same or similar functionality ?

Reply:
- Don't keep calling mapAttrsToList
  mapAttrsToList is `attrs: map fn (attrNames attrs)`.
  You should store attrNames in a variable and keep reusing that

- lib.remove null -> concatMap
  Nix has an optimisation for empty lists

The code below is a good simple example:

``` nix

# Check that problem has a message (required field)
++ lib.pipe meta.problems [
  (lib.mapAttrsToList (    name: problem:
    if problem ? message then
      null
    else
      "key 'meta.problems.${name}.message' is missing"
    ))
    (lib.remove null)
]
```

->

``` nix
let
  # Stick this in the outermost scope where it makes sense
  problemNames = attrNames meta.problems;
in
  # Nix has an optimisation for returning empty lists
  concatMap (name: if meta.problems.${name} ? message then [] else [ "key 'meta.problems.${name}.message' is missing" ])
```

Reply:
And just in general: Read the lib implementations you're using here and understand that Nix is not very optimised. 
Sadly we don't have good guidance on how to write performant Nix code.
`NIX_SHOW_STATS=1` is your friend.